Fix check_ssl_connection() function

kucharskim opened the pull request:
Function doesn't wait for `openssl s_client ...` to finish. It assumes that when the command is still running that is the successful condition. However the function should wait for exit code from the binary. We saw in production intermittent and very often `skale ssl upload` failures. This change should fix this problem and underlying race condition.

Related to:
- https://github.com/skalenetwork/node-cli/issues/868

kucharskim committed:
Fix check_ssl_connection() function

Function doesn't wait for `openssl s_client ...` to finish.
It assumes that when the command is still running that is the
successful condition. However the function should wait for
exit code from the binary. We saw in production intermittent
and very often `skale ssl upload` failures. This change
should fix this problem and underlying race condition.

kucharskim commented:
There sill maybe an issue in `node_cli/core/ssl/check.py` with other functions, which are on the client side, but I don't know how to test those code paths, so I prefer not to touch them.

kucharskim commented on node_cli/core/ssl/check.py:
```diff
@@ -202,7 +202,15 @@ def check_ssl_connection(host, port, silent=False):
-        code = dp.poll()
-        if code is not None:
+        for _ in range(10):
+            code = dp.poll()
```
I may use also `dp.wait(timeout)` instead of `dp.pool()` and a `for` loop.

badrogger replied:
Yes, please. It is a little more clear.

kucharskim replied:
Ok, I will do `dp.wait(timeout)` then.

badrogger commented on node_cli/core/ssl/check.py:
```diff
@@ -202,7 +202,15 @@ def check_ssl_connection(host, port, silent=False):
     expose_output = not silent
     with detached_subprocess(ssl_check_cmd, expose_output=expose_output) as dp:
         time.sleep(1)
```
This sleep is not needed anymore.

kucharskim committed:
Move to dp.wait() in check_ssl_connection()

Replace for loop and dp.poll() with more straightforward
dp.wait() with a timeout, as requested during diff review.

kucharskim commented on node_cli/core/ssl/check.py:
```diff
@@ -201,8 +201,15 @@ def check_ssl_connection(host, port, silent=False):
-            raise SSLHealthcheckError('OpenSSL connection verification failed')
+        timeout = 20
+        try:
+            dp.wait(timeout=timeout)
```
This consistently times out :/

Based on https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait I suspect:

> Note This will deadlock when using stdout=PIPE or stderr=PIPE and the child process generates enough output to a pipe such that it blocks waiting for the OS pipe buffer to accept more data. Use [Popen.communicate()](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate) when using pipes to avoid that.

kucharskim replied:
Reading now https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate

> [communicate()](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate) returns a tuple (stdout_data, stderr_data).

and looking how `def detached_subprocess()` is defined in `core/ssl/utils.py`, that it is `p.stdout.read()` and grabs the `output`, I am inclined to actually lave the loop and `dp.poll()` as it was.

kucharskim replied:
With `dp.wait()` it consistently throws `subprocess.TimeoutExpired` and `openssl s_client -connect 127.0.0.1:4536 -verify_return_error -verify 2` is not finishing, but it is running.

kucharskim replied:
Hm.. it times out for me now with `dp.poll()` as well, but it worked before :thinking:

kucharskim replied:
Ok, I think there were two problems and another fix was `stdin=subprocess.DEVNULL` in `utils.py`

kucharskim committed:
Read from /dev/null in detached_subprocess()

Redirect the child's standard input to subprocess.DEVNULL,
so it starts with no stdin attached. This prevents the OpenSSL
health-check process from reading from, or blocking on, the
parent's terminal or execution environment stdin stream.

kucharskim commented:
With all 3 commits, command `python main.py ssl upload -f -k <key.pem> -c <fullchain.pem>` finally works reliably on our machines.

kucharskim commented:
Tested my pull req interactively and via systemd unit file - both work fine.
